flink-ml-lib/src/main/java/org/apache/flink/ml/stats/chisqtest/ChiSqTest.java

```diff
@@ -0,0 +1,492 @@
+import org.apache.flink.ml.util.ParamUtils;
+import org.apache.flink.ml.util.ReadWriteUtils;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.BoundedOneInput;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.commons.math3.distribution.ChiSquaredDistribution;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Chi-square test of independence of variables in a contingency table. This Transformer computes
+ * the chi-square statistic,p-value,and dof(number of degrees of freedom) for every feature in the
+ * contingency table, which constructed from the `observed` for each categorical values. All label
+ * and feature values must be categorical.
+ *
+ * <p>See: http://en.wikipedia.org/wiki/Chi-squared_test.
+ */
+public class ChiSqTest implements AlgoOperator<ChiSqTest>, ChiSqTestParams<ChiSqTest> {
+
+    final String bcCategoricalMarginsKey = "bcCategoricalMarginsKey";
+    final String bcLabelMarginsKey = "bcLabelMarginsKey";
+
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+
+    public ChiSqTest() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+
+        final String[] inputCols = getInputCols();
+        String labelCol = getLabelCol();
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        SingleOutputStreamOperator<Tuple3<String, Object, Object>> colAndFeatureAndLabel =
+                tEnv.toDataStream(inputs[0])
+                        .flatMap(new ExtractColAndFeatureAndLabel(inputCols, labelCol));
+
+        // compute the observed frequencies
+        DataStream<Tuple4<String, Object, Object, Long>> observedFreq =
+                DataStreamUtils.mapPartition(
+                        colAndFeatureAndLabel.keyBy(Tuple3::hashCode),
+                        new GenerateObservedFrequencies());
+
+        SingleOutputStreamOperator<Tuple4<String, Object, Object, Long>> filledObservedFreq =
```

Comment:
How about we just compute the `distinct labels` and postpone the `fill` operation to Line#199, e.g., `DataStream<Tuple3<String, Double, Integer>> categoricalStatistics =...`?

Using `parallellism=1` for computing all data is not efficient usually.

Reply:
> How about we just compute the `distinct labels` and postpone the `fill` operation to Line#199, e.g., `DataStream<Tuple3<String, Double, Integer>> categoricalStatistics =...`?
> 
> Using `parallellism=1` for computing all data is not efficient usually.

Hi @zhipeng93 , thanks for the comments.If we need to adjust this method block,colud consider the following?

1. Typically, developers use discretized data as input,therefore,`GenerateObservedFrequencies `will reduce large amounts of input data before calling `FillZeroFunc`.

2. `FilledObservedFreq `represents the contingency table ,maybe it can be output as a result in the future? Refer to pandas `crosstab `in Python.

Reply:
Thanks for the explanation.

For 1: I agree that `GenerateObservedFrequencies` will reduce most of the input data in most cases, but there exist cases that there are many `zeros` in one column, e.g, there are many different labels but data is sparse. 
However, postponing the `FillZeroFunc` should always outperforms the current implementation.

For 2, I think we will not output `FilledObservedFreq` in `ChiSqTest`. We should be careful about the behavior of a well-known algorithm --- `ChiSqTest` does  not output the contingency table. If we need to output the contingency table, we may have a new algorithm for that.

Reply:
I think this is hard to avoid `parallellism=1`, because the cardinality of label is unpredictable,it will eventually aggregate on a node to distinguish the label.However, we can try to reduce the amount of data on the `Map`side.